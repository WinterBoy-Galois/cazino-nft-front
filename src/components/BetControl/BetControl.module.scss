@import '../../styles/shared';

$background-color: $control-background;
$label-color: $control-label;
$border-radius: 8px;
$border-width: 2px;
$border-focus-color: #fec41c;
$border-hover-color: #80a2cc;

.container {
  width: 100%;
  display: flex;
  flex-direction: column;
  background-color: $background-color;
  border-radius: $border-radius;
  padding: 5px 10px;
  border: $border-width solid $background-color;
  transition: all 0.2s ease-in;

  &__value {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    align-items: center;
  }

  label {
    text-transform: uppercase;
    display: block;
    font-size: 14px;
    line-height: 19px;
    color: $label-color;
    user-select: none;
    margin: 0;
  }

  input {
    color: $white;
    background-color: $background-color;
    margin: 0;
    padding: 0;
    font-size: 18px;
    line-height: 18px;
    border: 0;
    width: 100%;

    &:disabled {
      -webkit-text-fill-color: $white;
      opacity: 1;
    }
  }

  &:hover {
    border: $border-width solid $border-hover-color;
  }

  &:focus-within {
    border: $border-width solid $border-focus-color;
  }
}

.readonly {
  > * {
    cursor: pointer;
  }
}
