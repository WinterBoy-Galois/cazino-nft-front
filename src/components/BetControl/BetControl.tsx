import React, { useEffect, useState } from 'react';
import styles from './BetControl.module.scss';
import clsx from 'clsx';
import { isValid } from './lib/util';
import Icon from './components/Icon';
import { IconType } from './components/Icon/Icon';

interface IProps {
  label?: string;
  value?: number;
  className?: string;
  onChange?: (value: number) => void;
  min?: number;
  max?: number;
  icon?: IconType;
  decimalPlaces?: number;
  readonly?: boolean;
  onClick?: () => void;
}

const BetControl: React.FC<IProps> = ({
  label,
  value: defaultValue = 0.0,
  className,
  onChange,
  min = 0,
  max = 100,
  icon,
  decimalPlaces = 2,
  readonly,
  onClick,
}) => {
  const formatValue = (v: number) => v.toFixed(decimalPlaces);
  const [value, setValue] = useState(defaultValue);
  const [internalValue, setInternalValue] = useState(formatValue(defaultValue));
  const [editing, setEditing] = useState(false);

  useEffect(() => {
    if (!editing) {
      setValue(defaultValue);
      setInternalValue(formatValue(defaultValue));
    }
  }, [defaultValue]);

  const handleChange = (newValue: string) => {
    setEditing(true);

    setInternalValue(newValue);

    if (isValid(newValue, +min.toFixed(3), +max.toFixed(3))) {
      setValue(+newValue);
      onChange && onChange(+newValue);
    }
  };

  const handleBlur = () => {
    let newValue: number;

    if (min && +internalValue < min) {
      newValue = min;
    } else if (max && +internalValue > max) {
      newValue = max;
    } else {
      newValue = value;
    }

    setInternalValue(formatValue(newValue));
    setValue(+newValue);
    onChange && onChange(+newValue);

    setEditing(false);
  };

  return (
    <div
      className={clsx(styles.container, className, readonly && styles.readonly)}
      onClick={onClick}
    >
      <label htmlFor={label}>{label}</label>
      <div className={clsx(styles.container__value, readonly && styles.readonly)}>
        <Icon icon={icon} />
        <input
          id={label}
          type="text"
          inputMode="numeric"
          pattern="[0-9]*"
          value={internalValue}
          autoComplete="off"
          onChange={e => handleChange(e.target.value)}
          data-testid="value"
          onBlur={handleBlur}
          disabled={readonly}
        />
      </div>
    </div>
  );
};

export default BetControl;
