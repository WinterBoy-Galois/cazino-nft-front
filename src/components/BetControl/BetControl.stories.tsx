import React from 'react';
import { text, number, select } from '@storybook/addon-knobs';

import BetControl from '.';

export default {
  title: 'Components/BetControls/BetControl',
  component: BetControl,
};

const data = {
  label: 'Probability',
  value: 31.76,
  min: 0.000001,
  max: 0.1,
};

export const Default = () => (
  <BetControl
    label={text('Label', data.label)}
    value={number('Value', data.value)}
    min={number('Min', data.min)}
    max={number('Max', data.max)}
    icon={select(
      'Icon',
      { Probability: 'PROBABILITY', Multiplier: 'MULTIPLIER', OverUnder: 'OVER_UNDER' },
      'PROBABILITY'
    )}
  />
);
