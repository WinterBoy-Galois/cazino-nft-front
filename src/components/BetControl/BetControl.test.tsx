import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import BetControl from '.';

describe('BetControl', () => {
  it('should render', () => {
    // Arrange

    // Act
    const { container } = render(<BetControl />);

    // Assert
    expect(container).toMatchSnapshot();
  });

  it('should display 5', () => {
    // Arrange
    render(<BetControl value={5} decimalPlaces={0} />);

    // Act

    // Assert
    expect(screen.getByTestId('value')).toHaveValue('5');
  });

  xit('should display 1 for min 1', () => {
    // Arrange
    render(<BetControl value={0} min={1} />);

    // Act

    // Assert
    expect(screen.getByTestId('value')).toHaveValue('1');
  });

  xit('should display 5 for max 5', () => {
    // Arrange
    render(<BetControl value={10} max={5} />);

    // Act

    // Assert
    expect(screen.getByTestId('value')).toHaveValue('5');
  });

  it('should not call onChange value < min', () => {
    // Arrange
    const changeValue = jest.fn();
    render(<BetControl value={1} min={1} onChange={changeValue} />);

    // Act
    fireEvent.change(screen.getByTestId('value'), { target: { value: -5 } });

    // Assert
    expect(changeValue).toHaveBeenCalledTimes(0);
  });

  it('should call onChange value = min', () => {
    // Arrange
    const changeValue = jest.fn();
    render(<BetControl value={2} min={1} onChange={changeValue} />);

    // Act
    fireEvent.change(screen.getByTestId('value'), { target: { value: 1 } });

    // Assert
    expect(changeValue).toHaveBeenCalledTimes(1);
  });

  it('should call onChange value > min', () => {
    // Arrange
    const changeValue = jest.fn();
    render(<BetControl value={1} min={1} onChange={changeValue} />);

    // Act
    fireEvent.change(screen.getByTestId('value'), { target: { value: 3 } });

    // Assert
    expect(changeValue).toHaveBeenCalledTimes(1);
  });

  it('should call onChange value = max', () => {
    // Arrange
    const changeValue = jest.fn();
    render(<BetControl value={1} max={5} onChange={changeValue} />);

    // Act
    fireEvent.change(screen.getByTestId('value'), { target: { value: 5 } });

    // Assert
    expect(changeValue).toHaveBeenCalledTimes(1);
  });

  it('should call onChange value < max', () => {
    // Arrange
    const changeValue = jest.fn();
    render(<BetControl value={1} max={5} onChange={changeValue} />);

    // Act
    fireEvent.change(screen.getByTestId('value'), { target: { value: 4 } });

    // Assert
    expect(changeValue).toHaveBeenCalledTimes(1);
  });

  it('should not call onChange value > max', () => {
    // Arrange
    const changeValue = jest.fn();
    render(<BetControl value={1} max={5} onChange={changeValue} />);

    // Act
    fireEvent.change(screen.getByTestId('value'), { target: { value: 6 } });

    // Assert
    expect(changeValue).toHaveBeenCalledTimes(0);
  });

  it('should change quantity to 10', () => {
    // Arrange
    const container = render(<BetControl value={5} decimalPlaces={0} />);
    const input = container.getByTestId('value');

    // Act
    fireEvent.change(input, { target: { value: 10 } });

    // Assert
    expect((input as HTMLInputElement).value).toBe('10');
  });

  it('should revert to max 8', () => {
    // Arrange
    const container = render(<BetControl value={5} max={8} decimalPlaces={0} />);
    const input = container.getByTestId('value');

    // Act
    input.focus();
    fireEvent.change(input, { target: { value: 10 } });
    input.blur();

    // Assert
    expect((input as HTMLInputElement).value).toBe('8');
  });

  it('should revert to recent value', () => {
    // Arrange
    const container = render(<BetControl value={5} decimalPlaces={0} />);
    const input = container.getByTestId('value');

    // Act
    input.focus();
    fireEvent.change(input, { target: { value: 'test' } });
    input.blur();

    // Assert
    expect((input as HTMLInputElement).value).toBe('5');
  });
});
