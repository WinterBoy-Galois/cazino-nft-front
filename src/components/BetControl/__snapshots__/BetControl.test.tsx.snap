// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BetControl should render 1`] = `
<div>
  <div
    class="container"
  >
    <label />
    <div
      class="container__value"
    >
      <svg
        class="fill icon"
        viewBox="0 0 15.82 15.82"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          data-name="Layer 2"
        >
          <g
            data-name="Layer 1"
          >
            <path
              d="M4.5 4.5a2.64 2.64 0 10-3.73 0 2.64 2.64 0 003.73 0zm10.55 6.82a2.64 2.64 0 100 3.73 2.64 2.64 0 000-3.73zM15 1.78l-1-.93a1.33 1.33 0 00-1.87 0L.85 12.17a1.32 1.32 0 000 1.83l.93.93a1.33 1.33 0 001.87 0L15 3.65a1.33 1.33 0 000-1.87z"
              data-name="Icon awesome-percentage"
            />
          </g>
        </g>
      </svg>
      <input
        autocomplete="off"
        data-testid="value"
        inputmode="numeric"
        pattern="[0-9]*"
        type="text"
        value="0.00"
      />
    </div>
  </div>
</div>
`;
